tmol/score/cartbonded/potentials/cartbonded_pose_score.impl.hh

```diff
@@ -0,0 +1,367 @@
+
+  auto func = ([=] TMOL_DEVICE_FUNC(
+                   int pose_index, int block_index, int subgraph_index) {
+    Real score = 0;
+
+    int block_type = pose_stack_block_type[pose_index][block_index];
+    auto pose_coords = coords[pose_index];
+    int block_coord_offset =
+        pose_stack_block_coord_offset[pose_index][block_index];
+    int subgraph_offset = cart_subgraph_offsets[block_type];
+    int subgraph_offset_next = block_type + 1 == cart_subgraph_offsets.size(0)
+                                   ? cart_subgraphs.size(0)
+                                   : cart_subgraph_offsets[block_type + 1];
+    subgraph_index += subgraph_offset;
+
+    auto score_subgraph =
+        ([&] TMOL_DEVICE_FUNC(Vec<Int, 4> atoms, Int param_index) {
+          Vec<Real, 7> params = hash_values[param_index];
+
+          Vec<Real, 3> atom1 = pose_coords[atoms[0]];
+          Vec<Real, 3> atom2 = pose_coords[atoms[1]];
+
+          subgraph_type type = get_subgraph_type(atoms);
+
+          int score_type = params[0];
+
+          // Real score;
+          switch (type) {
+            case subgraph_type::length: {
+              auto eval = cblength_V_dV(atom1, atom2, params[2], params[1]);
+              accumulate_result<Real, Int, 2, D>(
+                  eval,
+                  atoms.head(2),
+                  V[score_type][pose_index],
+                  dV_dx[score_type][pose_index]);
+
+              break;
+            }
+            case subgraph_type::angle: {
+              Vec<Real, 3> atom3 = pose_coords[atoms[2]];
+              auto eval =
+                  cbangle_V_dV(atom1, atom2, atom3, params[2], params[1]);
+              accumulate_result<Real, Int, 3, D>(
+                  eval,
+                  atoms.head(3),
+                  V[score_type][pose_index],
+                  dV_dx[score_type][pose_index]);
+
+              break;
+            }
+            case subgraph_type::torsion: {
+              Vec<Real, 3> atom3 = pose_coords[atoms[2]];
+              Vec<Real, 3> atom4 = pose_coords[atoms[3]];
+              auto eval = cbtorsion_V_dV(
+                  atom1,
+                  atom2,
+                  atom3,
+                  atom4,
+                  params[1],
+                  params[2],
+                  params[3],
+                  params[4],
+                  params[5],
+                  params[6]);
+              accumulate_result<Real, Int, 4, D>(
+                  eval,
+                  atoms.head(4),
+                  V[score_type][pose_index],
+                  dV_dx[score_type][pose_index]);
+
+              break;
+            }
+          }
+        });
+
+    if (subgraph_index >= subgraph_offset_next) {
+      if (subgraph_index >= subgraph_offset_next + NUM_INTER_RES_PATHS) return;
+
+      subgraph_index -= subgraph_offset_next;
+
+      // Iterate over each connection in this block
+      for (int i = 0; i < pose_stack_inter_block_connections.size(2); i++) {
```

Comment:
I'll go ahead and add a comment, but basically it is because atom order matters (Cα–C'–N–Cα is different from Cα–N–C'–Cα), and we do not check the reverse path for inter-residue paths, only for intra.